File: Training/dualpal.cpp
/*
 ID: 
 PROG: dualpal
 LANG: C++
 */
//NAME          :   
//GROUP         :   Krypton
//LAST MODIFIED :   24 December 2017
//PROBLEM ID    :   USACO Dual Palindromes
//DESCRIPTION   :   Finds all numbers greater than a certain number S that are palindromes in 2 or more in the number bases 2-10
//SOURCES/HELP  :   
#include <iostream>
#include <cmath>
#include <string>
#include <fstream>
using namespace std;

//user defined function to convert base 10 numbers to any base 2-36
string baseConverter(int num10, int base)
{
    //declare and initialize variables
    int digit, powerNum = 0;
    string numB = "";
        //constant list has all digits for bases 2-36
    const string BASE_DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    int maxPower = 1;
    
    //loop to determine the first power greater than the base 10 number
    for (; maxPower <= num10; powerNum++)
    {
        maxPower = pow(base, powerNum);
    }
    
    //subtracts 1 from the exponent value of the first power greater than the base 10 number
    powerNum--;
    
    //loops while the exponent value is greater than 0
    while (powerNum > 0)
    {
        //decrements the power and exponent value
        maxPower /= base;
        powerNum--;
    
        //determines the highest integer multiple of the current highest power
        digit = floor(num10/maxPower);
        
        //subtracts the amount being added to the final number
        num10 -= maxPower*digit;
        
        //adds the digit to the number using the conversion list
        numB += BASE_DIGITS[digit];
        
    }
	//returns the converted number
    return numB;
}

//function to check if a string is a palindrome
bool checkPal(string number)
{
	//declare and initialize variables
		//find length of number
	int squareLength = number.length();
		//assumes number is not a palindrome
	bool isPal = false;
		//temporary variable for converted number
	string tempString = number;
		
	//runs for the floor of length divided by 2 to check both ends at the same time
	for (int j = 0; j < floor(squareLength/2); j++)
	{
		//checks if the numbers opposite of each other in the number are equal
		if (tempString[j] == tempString[squareLength-j-1])
		{
			//says that it is a palindrome
			isPal = true;
		}
		else
		{
			//say that it is not a palindrome
			isPal = false;
			
			//exits loop
			break;
		}
	}
	
	//checks if the number's length is one
	if (squareLength == 1)
	{
		isPal = true;
	}
	return isPal;
}

//main function
int main() 
{
	//declare variables
	int startNumber, amountOfNumbers, numBases, currentNum;
	bool isPal;
	
	//open input file
	ifstream dualpalin;
	dualpalin.open("dualpal.in");
	
	//read in amount of numbers needed and the starting number
	dualpalin >> amountOfNumbers >> startNumber;
	
	//close input file
	dualpalin.close();
	
	//open output file
	ofstream dualpalout;
	dualpalout.open("dualpal.out");
	
	//initializes the first number to check
	currentNum = startNumber + 1;
	
	//initialize counter variable for loop
	int counter = 0;
	
	//loops to get the correct amount of numbers
	while (counter < amountOfNumbers)
	{
		//initialize how many bases each number is a palindrome in
		numBases = 0;
		
		//loops for bases 2-10
		for (int j = 2; j <= 10; j++)
		{
			//exits loops if the number is palindromic in 2 or more bases
			if (numBases >= 2)
			{
				break;
			}
			
			//calls functions and increments the palindromic base counter if the converted number is a palindrome
			isPal = checkPal(baseConverter(currentNum, j));
			if (isPal)
			{
				numBases++;
			}
		}
		
		//checks if the number is a dual palindrome
		if (numBases >= 2)
		{
			//outputs number
			dualpalout << currentNum << endl;
			
			//increments the counter for the amount of numbers
			counter++;
		}
		
		//advances to next number
		currentNum++;
	}
	
	//close output file
	dualpalout.close();
	
	//exit program
	return 0;
}

File: Training/dualpal.java
/*
 ID: 
 PROG: dualpal
 LANG: JAVA
 */
//NAME          :   
//GROUP         :   APCS
//LAST MODIFIED :   6 November 2018
//PROBLEM ID    :   USACO Dual Palindromes
//DESCRIPTION   :   Finds all numbers greater than a certain number S that are palindromes in 2 or more in the number bases 2-10
//SOURCES/HELP  :   
import java.io.*;
import java.lang.*;
import java.util.StringTokenizer;
class dualpal
{
  //main function
  public static void main(String[] args) throws IOException
  {
    //file input
    BufferedReader in = new BufferedReader(new FileReader("dualpal.in"));
    StringTokenizer tokenizer = new StringTokenizer(in.readLine());

    //file output
    PrintWriter out = new PrintWriter(new BufferedWriter(new FileWriter("dualpal.out")));

    //declare variables
    int numsNeeded, startNum, currentNum, basesCorrect;

    //input and initialize variables
    numsNeeded = Integer.parseInt(tokenizer.nextToken());
    startNum = Integer.parseInt(tokenizer.nextToken());
    currentNum = startNum+1;

    //System.out.println(numsNeeded + " " + startNum);

    //loop until enough numbers are found
    while (numsNeeded > 0)
    {
      //initialize variable
      basesCorrect = 0;

      //loop through all bases 2-10
      for (int base = 2; base <= 10; base++)
      {
        //check if the number in the base is a palindrome
        if (palChecker(baseConvert(currentNum, base)))
        {
          basesCorrect++;
        }

        //case when the number is a palindrome in 2 or more bases
        if (basesCorrect == 2)
        {
          //decrement the amount of numbers needed and output
          numsNeeded--;
          out.println(currentNum);
          break;
        }
      }

      //move to next number
      currentNum++;
    }

    //exit program
    in.close();
    out.close();
    return;
  }

  //convert a base 10 number to any base 2-36
  public static String baseConvert(int dec, int base)
  {
    //declare string of values to use
    final String SYMBOLS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    String result = "";

    while (dec > 0)
    {
      //add the next digit
      result = SYMBOLS.charAt(dec%base) + result;

      //divide the base 10 number by the current base 
      dec /= base;
    }
    
    //return the converted number
    return result;
  }

  //function to check if a string is a palindrome
  public static boolean palChecker(String palindrome)
  {
    for (int i = 0; i < palindrome.length()/2; i++)
    {
      if (palindrome.charAt(i) != palindrome.charAt(palindrome.length()-i-1))
      {
        return false;
      }
    }
    return true;
  }
}